src/bitcoind.cpp

```diff
@@ -28,6 +29,79 @@
 const std::function<std::string(const char*)> G_TRANSLATION_FUN = nullptr;
 UrlDecodeFn* const URL_DECODE = urlDecode;
 
+#if HAVE_DECL_FORK
+
+/** Custom implementation of daemon(). This implements the same order of operations as glibc.
+ * Opens a pipe to the child process to be able to wait for an event to occur.
+ *
+ * @returns 0 if succesful, and in child process.
+ *          >0 if succesful, and in parent process.
+ *          -1 in case of error (in parent process).
+ *
+ *          In case of success, endpoint will be one end of a pipe from the child to parent process,
+ *          which can be used with TokenWrite (in the child) or TokenRead (in the parent).
+ */
+int fork_daemon(bool nochdir, bool noclose, TokenPipeEnd &endpoint)
+{
+    // communication pipe with child process
+    std::optional<TokenPipe> umbilical = TokenPipe::create();
+    if (!umbilical) {
+        return -1; // pipe or pipe2 failed.
+    }
+
+    int pid = fork();
+    if (pid < 0) {
+        return -1; // fork failed.
+    }
+    if (pid != 0) {
+        // Parent process gets read end, closes write end.
+        endpoint = umbilical->ReadEnd();
+        umbilical->WriteEnd();
+
+        int status = endpoint.TokenRead();
+        if (status != 0) { // Something went wrong while setting up child process.
+            endpoint.Close();
+            return -1;
+        }
+
+        return pid;
+    }
+    // Child process gets write end, closes read end.
+    endpoint = umbilical->WriteEnd();
+    umbilical->ReadEnd();
+
+#if HAVE_DECL_SETSID
+    if (setsid() < 0) {
+        exit(1); // setsid failed.
+    }
+#endif
+
+    if (!nochdir) {
+        if (chdir("/") != 0) {
+            exit(1); // chdir failed.
+        }
+    }
+    if (!noclose) {
+        // Open /dev/null, and clone it into STDIN, STDOUT and STDERR to detach
+        // from terminal.
+        int fd = open("/dev/null", O_RDWR);
+        if (fd >= 0) {
+            bool err = dup2(fd, STDIN_FILENO) < 0 || dup2(fd, STDOUT_FILENO) < 0 || dup2(fd, STDERR_FILENO) < 0;
+            close(fd);
+            if (err) {
+                exit(1); // dup2 failed.
+            }
+            close(fd);
```

Comment:
The standard reason to compare it to 2 is in case you start the executable with stdin/stdout/stderr already closed -- in that case `fd` will be assigned to 0/1/2 because that's the first free fd, `dup2(fd,stdin)` (or out or err) will evaluate to `dup2(0,0)` and be a no-op, and you'll then call `close(0)` which will leave you without a dummied stdin (or out or err) which will mean some later fd you open might overlap with stdin (or out or err). Not sure that things wouldn't already be broken in those cases though -- TokenPipe would probably grab the free fds in that case, and then dup2 would close the TokenPipe fd.

Reply:
Hmm good point. I did not think about that case. It's really unconventional to start binaries with stdin/stdout/stderr closed (the graceful way would be to redirect them to `/dev/null` and I'm not sure we handle this in other places.